subxt/src/transaction.rs

```diff
@@ -440,60 +434,73 @@ impl<'client, T: Config, E: Decode> TransactionInBlock<'client, T, E> {
     /// Return the hash of the extrinsic.
     pub fn extrinsic_hash(&self) -> T::Hash {
         self.ext_hash
     }
 
-    /// Return a slice of the returned events.
-    pub fn as_slice(&self) -> &[crate::RawEvent] {
+    /// Return all of the events in the block that the transaction made it into.
+    pub fn all_events_in_block(&self) -> &events::Events<'client, T, Evs> {
         &self.events
     }
 
-    /// Find all of the events matching the event type provided as a generic parameter. This
-    /// will return an error if a matching event is found but cannot be properly decoded.
-    pub fn find_events<Ev: crate::Event>(&self) -> Result<Vec<Ev>, BasicError> {
-        self.events
-            .iter()
-            .filter_map(|e| e.as_event::<Ev>().map_err(Into::into).transpose())
-            .collect()
+    /// Iterate over the statically decoded events associated with this transaction. Works
+    /// in the same way that [`events::Events::iter()`] does, with the
+    /// exception that it filters out events not related to the submitted extrinsic.
+    pub fn iter(
+        &self,
+    ) -> impl Iterator<Item = Result<EventDetails<Evs>, BasicError>> + '_ {
+        self.events.iter().filter(|ev| {
+            ev.as_ref()
+                .map(|ev| ev.phase == Phase::ApplyExtrinsic(self.ext_idx))
+                .unwrap_or(true) // Keep any errors
+        })
     }
 
-    /// Find the first event that matches the event type provided as a generic parameter. This
-    /// will return an error if a matching event is found but cannot be properly decoded.
-    ///
-    /// Use [`TransactionEvents::find_events`], or iterate over [`TransactionEvents`] yourself
-    /// if you'd like to handle multiple events of the same type.
-    pub fn find_first_event<Ev: crate::Event>(&self) -> Result<Option<Ev>, BasicError> {
-        self.events
-            .iter()
-            .filter_map(|e| e.as_event::<Ev>().transpose())
-            .next()
-            .transpose()
-            .map_err(Into::into)
+    /// Iterate over all of the raw events associated with this transaction. Works
+    /// in the same way that [`events::Events::iter_raw()`] does, with the
+    /// exception that it filters out events not related to the submitted extrinsic.
+    pub fn iter_raw(
+        &self,
+    ) -> impl Iterator<Item = Result<RawEventDetails, BasicError>> + '_ {
+        self.events.iter_raw().filter(|ev| {
+            ev.as_ref()
+                .map(|ev| ev.phase == Phase::ApplyExtrinsic(self.ext_idx))
+                .unwrap_or(true) // Keep any errors.
+        })
     }
 
-    /// Find an event. Returns true if it was found.
-    pub fn has_event<Ev: crate::Event>(&self) -> Result<bool, BasicError> {
-        Ok(self.find_first_event::<Ev>()?.is_some())
+    /// Find all of the transaction events matching the event type provided as a generic parameter.
+    /// This will return an error if a matching event is found but cannot be properly decoded.
```

Comment:
…and interrupt the iterator yes? Also: worth mentioning that this uses dynamic decoding?

Reply:
I'll point all of these methods at the corresponding versions on `Events`, since they work identically with the exception that they ignore events not related to the submitted extrinsic.